filebeat/input/filestream/internal/input-logfile/manager.go

```diff
@@ -166,34 +167,53 @@ func (cim *InputManager) Create(config *common.Config) (input.Input, error) {
 		return nil, errNoInputRunner
 	}
 
+	sourceIdentifier := newSourceIdentifier(cim.Type, settings.ID)
+	pStore := cim.getRetainedStore()
+	defer pStore.Release()
+	prospectorStore := newSourceStore(pStore, sourceIdentifier)
+	err = prospector.Init(prospectorStore)
+	if err != nil {
+		return nil, err
+	}
+
 	return &managedInput{
-		manager:      cim,
-		prospector:   prospector,
-		harvester:    harvester,
-		cleanTimeout: settings.CleanTimeout,
+		manager:          cim,
+		userID:           settings.ID,
+		prospector:       prospector,
+		harvester:        harvester,
+		sourceIdentifier: sourceIdentifier,
+		cleanTimeout:     settings.CleanTimeout,
 	}, nil
 }
 
-// Lock locks a key for exclusive access and returns an resource that can be used to modify
-// the cursor state and unlock the key.
-func (cim *InputManager) lock(ctx input.Context, key string) (*resource, error) {
-	resource := cim.store.Get(key)
-	err := lockResource(ctx.Logger, resource, ctx.Cancelation)
-	if err != nil {
-		resource.Release()
-		return nil, err
-	}
-	return resource, nil
+func (cim *InputManager) getRetainedStore() *store {
+	store := cim.store
+	store.Retain()
+	return store
 }
 
-func lockResource(log *logp.Logger, resource *resource, canceler input.Canceler) error {
-	if !resource.lock.TryLock() {
-		log.Infof("Resource '%v' currently in use, waiting...", resource.key)
-		err := resource.lock.LockContext(canceler)
-		if err != nil {
-			log.Infof("Input for resource '%v' has been stopped while waiting", resource.key)
-			return err
-		}
+type sourceIdentifier struct {
+	prefix           string
+	configuredUserID bool
+}
+
+func newSourceIdentifier(pluginName, userID string) *sourceIdentifier {
+	inputPrefix := pluginName
+	configuredUserID := false
+	if userID != "" {
+		inputPrefix = inputPrefix + "-" + userID
+		configuredUserID = true
 	}
-	return nil
+	return &sourceIdentifier{
+		prefix:           inputPrefix,
+		configuredUserID: configuredUserID,
+	}
+}
+
+func (i *sourceIdentifier) ID(s Source) string {
+	return i.prefix + "::" + s.Name()
+}
+
+func (i *sourceIdentifier) MatchesInput(id string) bool {
+	return strings.Contains(id, i.prefix)
```

Comment:
1. use 'StartsWith'.

2. the way `prefix` is constructed we can have false positives here. for an input without ID the prefix will be `filestream` and for and input with ID the prefix will be `filestream-<id>`. The former case is always a prefix of the later one. If you start to include the `::` into the prefix, then you have a tie breaker, because `filstream::` is no prefix of  filestream-<id>::`. This is some dependency between 'sourceIdentifier.prefix` construction and matching and should be documented in the constructor why `prefix` is build the way it is.

Reply:
Wrong language.. `strings.HasPrefix(id, i.prefix)` is what we should use.

Reply:
:facepalm: